### PR TITLE
Spaces: Prevent same default sync folder for multiple accounts

### PR DESCRIPTION
**Pull request by erikjv**
Fixes: https://github.com/owncloud/client/issues/11204

### CODE REVIEW

File: src/libsync/filesystem.cpp [hunk 1]
```diff
@@ -198,5 +203,96 @@ bool FileSystem::getInode(const QString &filename, quint64 *inode)
...
+    {
+        return ::removexattr(path, name);
+    }
+#endif // Q_OS_LINUX
+
+} // anonymous namespace
+
+static constexpr unsigned MaxValueSize = 1023; // This is without a terminating NUL character
+
+std::optional<QByteArray> FileSystem::Tags::get(const QString &path, const QString &key)
```
Comment (TheOneRing) [inline src/libsync/filesystem.cpp:229]: Get and set should return the same type (string/bytearray)

**Comment by erikjv** [comment]
To do: do the check for the spaces tag while checking folder paths, so recursively.

File: src/gui/guiutility.cpp [hunk 2]
```diff
@@ -85,3 +86,29 @@ QString Utility::vfsFreeSpaceActionText()
 {
     return QCoreApplication::translate("utility", "Free up local space");
 }
+
+static const QString dirTag = QStringLiteral("com.owncloud.spaces.app");
```
Comment (TheOneRing) [inline src/gui/guiutility.cpp:90]: please move to anonymous namespace and make it a function

File: src/gui/guiutility.h [hunk 3]
```diff
@@ -51,6 +51,10 @@ namespace Utility {
 
     QString socketApiSocketPath();
 
+    void setDirectoryTag(const QString &path);
```
Comment (TheOneRing) [inline src/gui/guiutility.h:54]: Please add a description.
Could you think of a more descriptive function name?

File: src/gui/guiutility.cpp [hunk 4]
```diff
@@ -85,3 +93,32 @@ QString Utility::vfsFreeSpaceActionText()
...
     return QCoreApplication::translate("utility", "Free up local space");
 }
+
+void Utility::setDirectoryTag(const QString &path)
+{
+    auto result = FileSystem::Tags::set(path, dirTag(), Theme::instance()->orgDomainName().toUtf8());
+    if (!result) {
+        qCWarning(lcGuiUtility) << QStringLiteral("Failed to set tag on '%1': %2").arg(path, result.error())
+#ifdef Q_OS_WIN
+                                << QStringLiteral("(filesystem %1)").arg(FileSystem::fileSystemForPath(path))
```
Comment (TheOneRing) [inline src/gui/guiutility.cpp:103]: Wouldn't that make more sense to check in FileSystem::Tags::set(? sry for not clarifying earlier

File: src/libsync/filesystem.cpp [hunk 5]
```diff
@@ -198,5 +207,100 @@ bool FileSystem::getInode(const QString &filename, quint64 *inode)
...
+        return file.readAll();
+    }
+#endif // Q_OS_MAC || Q_OS_LINUX
+
+    return {};
+}
+
+OCC::Result<void, QString> FileSystem::Tags::set(const QString &path, const QString &key, const QByteArray &value)
+{
+    OC_ASSERT(value.size() < MaxValueSize)
```
Comment (TheOneRing) [inline src/libsync/filesystem.cpp:257]: Should we error on implicit overrides? Or would that only make sense for setDirectoryTag?

**Review by TheOneRing — CHANGES_REQUESTED** [review]

File: src/gui/guiutility.h (hunk 3 above)
Comment (erikjv) [reply src/gui/guiutility.h:54]: what about `markDirectoryAsSyncRoot(..)`?

File: src/gui/guiutility.cpp (hunk 4 above)
Comment (erikjv) [reply src/gui/guiutility.cpp:103]: I kinda like doing the handling of the error outside of the "low level" function.

File: src/gui/guiutility.h (hunk 3 above)
Comment (TheOneRing) [reply src/gui/guiutility.h:54]: Better :)

**Comment by TheOneRing** [comment]
Please also add a changelog entry

File: src/gui/guiutility.h (hunk 3 above)
Comment (erikjv) [reply src/gui/guiutility.h]: done

File: src/libsync/filesystem.cpp (hunk 5 above)
Comment (erikjv) [reply src/libsync/filesystem.cpp:257]: Neither? This is when one sync app is removed and another one is installed. Then the new application can take over the old sync root. Or is this a bad idea?

Comment (TheOneRing) [reply src/libsync/filesystem.cpp:257]: The assert would only occur in dev setups.
For the above assert a Q_ASSERT should be sufficient.

I'd say a normal scenario would be that a sync root is properly removed, so we also removed the tag?
Maybe not an assert, but a general warning as the folder was not properly cleaned?

**Comment by TheOneRing** [comment]
Changelog is still missing

File: src/libsync/filesystem.cpp (hunk 5 above)
Comment (erikjv) [reply src/libsync/filesystem.cpp:257]: Done

**Review by TheOneRing — APPROVED** [review]

**Comment by TheOneRing** [comment]
Please fix the changelog:
> file changelog/unreleased/11204: entry type "Spaces" is invalid, valid types: map[Bugfix:2 Change:3 Enhancement:4 Security:1]

**Review by TheOneRing — APPROVED** [review]

**Comment by Salipa-Gurung** [comment]
Tested https://github.com/owncloud/client/issues/11204#issuecomment-1880467290

**Pull request merged**